src/OrbitQt/MainThreadExecutorImplTest.cpp

```diff
@@ -22,6 +22,93 @@ TEST(MainThreadExecutorImpl, Schedule) {
   EXPECT_EQ(app.exec(), 42);
 }
 
+TEST(MainThreadExecutorImpl, ScheduleAfterAllVoid) {
+  QCoreApplication app{argc, nullptr};
+
+  bool called = false;
+  auto executor = MainThreadExecutorImpl::Create();
+  auto future = executor->Schedule([&called]() { called = true; });
+  (void)executor->ScheduleAfter(future, []() { QCoreApplication::exit(42); });
+
+  EXPECT_EQ(app.exec(), 42);
+  EXPECT_TRUE(called);
+}
+
+TEST(MainThreadExecutorImpl, ScheduleAfterWithIntegerBetweenJobs) {
+  QCoreApplication app{argc, nullptr};
+
+  auto executor = MainThreadExecutorImpl::Create();
+  auto future = executor->Schedule([]() { return 42; });
+  (void)executor->ScheduleAfter(future, [](int val) { QCoreApplication::exit(val); });
+
+  EXPECT_EQ(app.exec(), 42);
+}
+
+TEST(MainThreadExecutorImpl, ScheduleAfterWithIntegerAsFinalResultAndBetweenJobs) {
+  QCoreApplication app{argc, nullptr};
+
+  auto executor = MainThreadExecutorImpl::Create();
+  auto future = executor->Schedule([]() { return 42; });
+  auto future2 = executor->ScheduleAfter(future, [](int val) {
+    QCoreApplication::exit(val);
+    return val + 42;
+  });
+
+  EXPECT_EQ(app.exec(), 42);
+  EXPECT_TRUE(future2.IsFinished());
+  EXPECT_EQ(future2.Get(), 42 + 42);
+}
+
+TEST(MainThreadExecutorImpl, ScheduleAfterWithIntegerOnlyAsFinalResult) {
+  QCoreApplication app{argc, nullptr};
+
+  bool called = false;
+  auto executor = MainThreadExecutorImpl::Create();
+  auto future = executor->Schedule([&called]() { called = true; });
+  auto future2 = executor->ScheduleAfter(future, []() {
+    QCoreApplication::exit(42);
+    return 42 + 42;
+  });
+
+  EXPECT_EQ(app.exec(), 42);
+  EXPECT_TRUE(future2.IsFinished());
+  EXPECT_EQ(future2.Get(), 42 + 42);
+  EXPECT_TRUE(called);
+}
+
+TEST(MainThreadExecutorImpl, ScheduleAfterMultipleContinuations) {
+  QCoreApplication app{argc, nullptr};
+
+  auto executor = MainThreadExecutorImpl::Create();
+  auto future = executor->Schedule([]() { return 42; });
+  auto future2 = executor->ScheduleAfter(future, [](int val) { return val + 42; });
+  auto future3 = executor->ScheduleAfter(future2, [](int val) { return val + 42; });
+  auto future4 =
+      executor->ScheduleAfter(future3, [](int val) { QCoreApplication::exit(val + 42); });
+
+  EXPECT_EQ(app.exec(), 4 * 42);
+}
+
+TEST(MainThreadExecutorImpl, ScheduleAfterWithExecutorOutOfScope) {
+  QCoreApplication app{argc, nullptr};
+
+  orbit_base::Promise<void> promise;
+  orbit_base::Future<void> future = promise.GetFuture();
+
+  auto executor = MainThreadExecutorImpl::Create();
+
+  bool called = false;
+  auto future2 = executor->ScheduleAfter(future, [&called]() { called = true; });
```

Comment:
ok, I see now. this is one case. But I still wonder if this has unintended consequences, if we do not guarantee execution of ScheduledAfter actions.

Reply:
This is very much intended behaviour. At some point I want to introduce the same behaviour for `Schedule`, but it's not that pressing there, since these actions are mostly executed right away. But let me quickly motivate why I think it is a good idea to tie the lifetime of actions to the lifetime of the executor:

We have a lot of Backgroundtasks that work like this:
```
void OrbitApp::Method() {
  thread_pool_->Schedule([this](){
    // do something expensive
    main_thread_executor_->Schedule([result]() {
      // call some member function of OrbitApp and process result
    });
  });
```

The thread pool action usually doesn't check whether `this` is still alive, nor if `main_thread_executor_` is still a valid pointer. 
The same applies to the main_thread_executor_ action. It also usually doesn't check that.

And we had already fixed crashes that occur when the application shuts down. There is also currently an open
bug: Orbit crashes sometimes when you close it while it's loading symbols.

There are two ways to solve that:
1. Make all actions check whether pointers are still valid. That's probably not feasible.
2. Tie the lifetime of actions to the objects they mutate. And that's what I'm trying here. The idea is to give `OrbitApp` (and also other objects if needed) their own instance of MainThreadExecutor. That means all actions running on this instance of MainThreadExecutor can assume that `OrbitApp` is still alive. And when `OrbitApp` gets destructed the local MainThreadExecutor dies first and no action will execute on dangling pointers.

Reply:
makes sense.